Add a basic Options modal

noisysocks opened the pull request:
Implements half of #10210.

Adds a basic Options modal:

![localhost_8888_wp-admin_post-new php screen recording 2](https://user-images.githubusercontent.com/612155/46160670-4473b600-c27b-11e8-9376-92324fda1916.png)

Not included in this PR is the ability to toggle "advanced panels" AKA metaboxes.

**To test:**

1. Open the editor
2. Open the _More_ drop-down menu in the top right of Gutenberg
3. Select  _Options_
4. Check that each setting works and that they're persisted when the page reloads

noisysocks commented on edit-post/store/reducer.js:
```diff
@@ -49,6 +51,20 @@ export const preferences = combineReducers( {
+
+		return state;
+	},
 	panels( state = PREFERENCES_DEFAULTS.panels, action ) {
```
Wondering if I should modify the existing `panels` state tree for our needs e.g. `panels: { 'post-status': { enabled: true, shown: false } }` instead of adding a whole new `disabledPanels`. And, if so, is it worth writing migration logic so that the shown/hidden panel state isn't lost when users update Gutenberg? @talldan

noisysocks replied:
Think I'll merge `disabledPanels` and `panels`. It doesn't matter that users upgrading Gutenberg will have to re-open a few panels here and there.

talldan commented:
Hey - had a test of this. Looks nice. 👍 

When toggling tips you I saw this, which looks a bit weird. I wonder whether the modal should be elevated over the tip. Fun with z-indexes 😄

<img width="549" alt="screen shot 2018-09-30 at 3 45 03 pm" src="https://user-images.githubusercontent.com/677833/46261949-bab92800-c4c8-11e8-96ab-8adf4629f3ab.png">

Confusingly, clicking outside of the tip dimisses the modal, even though the tip is on top.

Also, separately, the blank space at the bottom of the modal seems like it shouldn't be there. I imagine you'd only see that on larger screen sizes:

<img width="498" alt="screen shot 2018-09-30 at 3 59 31 pm" src="https://user-images.githubusercontent.com/677833/46262051-f86a8080-c4c9-11e8-81dd-76096da32a01.png">

noisysocks commented:
> When toggling tips you I saw this, which looks a bit weird. I wonder whether the modal should be elevated over the tip. Fun with z-indexes 😄

Hmm, but then what happens when we inevitably want to have a tip point to something in a modal? I might look at saving the Tips on/off setting when the modal is dismissed.

> Also, separately, the blank space at the bottom of the modal seems like it shouldn't be there. I imagine you'd only see that on larger screen sizes:

👌 will look into this! Thanks for the review.

talldan commented:
> Hmm, but then what happens when we inevitably want to have a tip point to something in a modal?

I also thought that, but then thought probably not worth worrying about until it happens.

> I might look at saving the Tips on/off setting when the modal is dismissed.

That might be a better user experience generally - definitely less noisy. Just for tips though or all options? Only drawback would be if the user tries to navigate away with unsaved changes to the checkboxes.

noisysocks commented on edit-post/components/options-modal/index.js:
```diff
@@ -0,0 +1,179 @@
+			isPublishSidebarEnabled: isPublishSidebarEnabled(),
+			areTipsEnabled: areTipsEnabled(),
+			isStatusPanelEnabled: isEditorSidebarPanelEnabled( 'post-status' ),
+			isCategoriesPanelEnabled: isEditorSidebarPanelEnabled( 'taxonomy-panel-category' ),
```
Users can define their own taxonomies so we'll need to add some weird dynamics here.

noisysocks committed:
Implements an Options modal

- Adds a basic Options modal to the More menu.
- Moves the 'Enable Pre-publish Checks' and 'Show Tips' toggles to the
  new modal.
- Allows the user to turn Document panels on or off.

noisysocks commented:
Hi! 👋 

@talldan: I've rebased this and updated it to handle custom taxonomies and set the 'Show Tips' setting when the modal is dismissed. Would you mind giving it another comprehensive review?

@gziolo: I did not add an API that lets developers disable block inspector panels to this PR. I'd rather that we tackled this in a future PR so that we can give that new code the attention that it deserves.

talldan reviewed (commented):
@noisysocks - This looks really good! In terms of UX, I only have a concern about the inconsistency with some options enabling immediately and others being deferred, but lets proceed see if anyone mentions it (probably unlikely).

I think I might have spotted a couple of things around deprecations, but let me know if I'm wrong.

gziolo commented on packages/edit-post/src/store/actions.js:
```diff
@@ -94,18 +94,49 @@ export function togglePublishSidebar() {
+ * @return {Object} Action object.
+*/
+export function toggleGeneralSidebarEditorPanel( panelName ) {
+	deprecated( 'toggleGeneralSidebarEditorPanel', {
```
Correct 👍

noisysocks replied:
👌 697c396f3

noisysocks replied:
👌 697c396f3

gziolo commented on packages/edit-post/src/store/reducer.js:
```diff
@@ -50,11 +50,29 @@ export const preferences = combineReducers( {
+
+			case 'TOGGLE_PANEL_OPENED': {
+				const { panelName } = action;
+				const isOpen = state[ panelName ] === true || get( state, [ panelName, 'opened' ], false );
```
Backward compatible 💯

noisysocks replied:
🙂

gziolo commented on packages/edit-post/src/store/selectors.js:
```diff
@@ -104,16 +104,31 @@ export function isPublishSidebarOpened( state ) {
- * @return {boolean} Whether the sidebar panel is open.
+ * @return {boolean} Whether or not the panel is enabled.
  */
-export function isEditorSidebarPanelOpened( state, panel ) {
```
Yes, we list them in docs.

noisysocks replied:
👌 f7b38cb9b

noisysocks replied:
👌 f7b38cb9b

gziolo commented on test/e2e/specs/nux.test.js:
```diff
@@ -103,12 +109,21 @@ describe( 'New User Experience (NUX)', () => {
-		// Click again to re-enable tips; they should appear.
-		await clickOnMoreMenuItem( 'Show Tips' );
+		// Open the Options modal.
+		await clickOnMoreMenuItem( 'Options' );
```
It's possible to generalize `togglePrePublishChecks` utility method to take the text name as param and put it inside the following string `//label[contains(text(), "Enable Tips")]` and thus simplify this code here and a few lines earlier.

noisysocks replied:
How's this? 1a4394918

gziolo commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+			</Section>
+			<Section title={ __( 'Document Panels' ) }>
+				<EnablePanelOption panelName="post-status" label={ __( 'Status & Visibility' ) } />
+				<PostTaxonomies
```
@talldan we can't do much about it, this can change depending on the website when a plugin registers their own taxonomy.

@noisysocks is this something that is already pre-cached by PHP, if not maybe we should consider doing it on PHP side to avoid network roundtrip. We do it already for a few essential API calls:

https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php#L1303-L1310

It looks like taxonomies is include, so it shouldn't be an issue here. We should double check.

noisysocks replied:
You're right that it taxonomies are already included in the list of API calls to cache.

When I test this locally it doesn't seem to be working, but since this issue also exists on `master` I will look it separately.

gziolo reviewed (commented):
I hope to do more in-depth review on Monday. However, don't wait for me if you think it's ready to go. We already pair-reviewed most of the code so this is definitely aligned with what I envisioned for this feature.

gziolo commented:
> I did not add an API that lets developers disable block inspector panels to this PR. I'd rather that we tackled this in a future PR so that we can give that new code the attention that it deserves.

I didn't expect that, I think your data related changes look solid, it should make this earlier in the future. I think that inspector panels won't be customizable until we start phase 2. It's too much work. Let's focus on backward compatibility.

gziolo commented:
I noticed 3 design issues which should be further discussed:

1. When all document options are disabled, it renders empty content, should we add a button which allows to enable some of the panels through the modal?
![screen shot 2018-10-15 at 11 21 51](https://user-images.githubusercontent.com/699132/46942026-b9613100-d06c-11e8-8aa2-19e534d7683a.png)
2. Options were moved to their own group which doesn't have a title, should it remain in one of the existing groups instead?
![screen shot 2018-10-15 at 11 22 25](https://user-images.githubusercontent.com/699132/46942027-b9613100-d06c-11e8-9f9f-c45ad06f6801.png)
3. You can toggle options by clicking checkbox or text, in my opinion we should expand it to the full width of the row
![screen shot 2018-10-15 at 11 23 56](https://user-images.githubusercontent.com/699132/46942088-deee3a80-d06c-11e8-858d-349f97845be2.png)

gziolo commented:
I can confirm that all functionalities work as expected. All options are properly saved and are persisted between page loads.

gziolo commented on packages/edit-post/src/components/options-modal/test/index.js:
```diff
@@ -0,0 +1,21 @@
+
+	it( 'should match snapshot when the modal is not active', () => {
+		const wrapper = shallow( <OptionsModal isModalActive={ false } /> );
+		expect( wrapper ).toMatchSnapshot();
```
Minor: It might be easier to read when you check in the test if it renders nothing instead. Snapshots are a bit indirect as you need to navigate to another file.

noisysocks replied:
👌 11f7c2d55

gziolo commented on packages/edit-post/src/store/selectors.js:
```diff
@@ -104,16 +104,31 @@ export function isPublishSidebarOpened( state ) {
+	const panels = getPreference( state, 'panels' );
+	return panels[ panelName ] === true || get( panels, [ panelName, 'opened' ], false );
 }
 
```
We should add back `isEditorSidebarPanelOpened` and alias `isEditorPanelOpened` with deprecation.

noisysocks replied:
👌 f7b38cb9b

gziolo reviewed (changes requested):
I left my comments. I would love to get some design related feedback before we proceed. The only code related concern I have is the missing deprecation for one of the selectors.

mtias commented:
> When all document options are disabled

We shouldn't allow "status and visibility" to be hidden.

gziolo commented:
I discussed backward compatibility with @mtias. We should respect values set in the Classic editor for the boxes. It can be retrieved using `get_hidden_meta_boxes` (https://github.com/WordPress/wordpress-develop/blob/6b366c6620fdd5960cedcdf80955966a715efa82/src/wp-admin/includes/screen.php#L151) function. It probably should be done as a follow-up task. We should only respect previously set values and then Gutenberg should take over control.

karmatosed commented:
> Hmm, but then what happens when we inevitably want to have a tip point to something in a modal?

> I also thought that, but then thought probably not worth worrying about until it happens.

Let's avoid having Tips showing if modal shows. We can iterate.

afercia commented:
When using only the keyboard and the modal closes, focus must return to the "invoking context", that is the control that opened the modal. This is an accessibility requirement for modals and it's how the Gutenberg modal is supposed to work. It has been built do have this behavior out of the box.

As the control that opened the modal disappears because the more menu is closed, there's a focus loss. Keyboard users, including screen reader users, will be forced to start keyboard navigation again from scratch from the document root. This is one of the worst experiences for these users, as the context they were operating in is completely lost, and should be avoided.

The only way I see (so far) to avoid the focus loss is to implement a fallback mechanism in the modal component, allowing to specify a fallback element where to move focus back to. The "More" button in the top bar is the closest element to the invoking context, it's not ideal but it's the only place I can think of.

tofumatt reviewed (commented):
This options modal would be the obvious place for Accessibility settings as well, so we should definitely make sure keyboard access, screenreader access, and click targets are really accessible so users can get to the options (we'd eventually put for accessibility) here.

noisysocks commented on packages/edit-post/src/components/options-modal/options.js:
```diff
@@ -0,0 +1,76 @@
+	}
+}
+
+export const EnablePublishSidebarOption = compose(
```
Yeah I'm pretty pleased with this!

noisysocks committed:
Add isEditorSidebarPanelOpened as a deprecated alias to isEditorPanelEnabled

noisysocks committed:
Document toggleGeneralSidebarEditorPanel deprecation

noisysocks committed:
Remove redundant WordPress dependencies header

noisysocks committed:
Rename classes from edit-post-options to edit-post-options-modal

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Internal dependencies
+ */
+import { EnablePublishSidebarOption, EnableTipsOption, EnablePanelOption } from './options';
```
I think it aids with comprehension to have these closely related components declared in the same source file. _One component per file_ is a nice guideline but in this case I think it would just annoy the reader to make them have to click around a lot. 

With regards to the names, I like that they currently describe the UI that's shown. For example, the `EnableTipsOption` is literally the _Enable Tips_ option:

<img width="440" alt="screen shot 2018-10-16 at 13 33 31" src="https://user-images.githubusercontent.com/612155/46989241-1978d580-d148-11e8-8471-9e58528911a7.png">

I'm happy to concede on any of this if you feel strongly about it! 🙂

gziolo replied:
Well, it depends on the reader. We aren't consistent at all in the codebase. Let's keep it as is in one file :)

We don't have strict rules for component names. I personally try to avoid using verbs at the beginning of the component as it implies action happening. I don't feel strongly about it, too. Your reasoning is also correct.

noisysocks replied:
Naming things, hey! 😅

noisysocks committed:
Move Section component to its own file

noisysocks committed:
Explain why we use DeferredOption

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	return (
+		<Modal
```
💃

noisysocks commented on packages/edit-post/src/components/options-modal/options.js:
```diff
@@ -0,0 +1,76 @@
+			onChange: ( isEnabled ) => ( isEnabled ? enableTips() : disableTips() ),
+		};
+	} )
+)( DeferredOption );
```
> We should document why we use a deferred option in here. I assume it's to ensure you don't see tips them when interacting with the modal.

That's correct. I've added an explanatory comment in 9ebb7bacd.

> It still raises the question why this needs to be a special case.

I don't want to compromise on what I think is the best UX which is that:

- Deselecting _Enable Pre-publish Checks_ immediately updates the _Publish..._ button to be _Publish_
- Toggling any of the _Document Panels_ options immediately updates the sidebar UI
- Selecting _Enable Tips_ does not display the NUX modal immediately. This is jarring since it interrupts the user's current task

gziolo replied:
This is helpful. It might prevent someone from changing it accidentally to immediate change.

noisysocks committed:
Remove toggleModal() in lieu of just closeModal()

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+	withDispatch( ( dispatch, { isModalActive } ) => {
+		const { openModal, closeModal } = dispatch( 'core/edit-post' );
+		return {
+			toggleModal: () => ( isModalActive ? closeModal() : openModal( MODAL_NAME ) ),
```
👌 bcb529a18

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+		<h2 className="edit-post-options__section-title">{ title }</h2>
+		{ children }
+	</section>
+);
```
👌 8a14afd2f

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+const MODAL_NAME = 'edit-post/options';
+
+const Section = ( { title, children } ) => (
+	<section className="edit-post-options__section">
```
👌 056e75b05

noisysocks commented on packages/edit-post/src/components/header/options-menu-item/index.js:
```diff
@@ -0,0 +1,31 @@
+ * WordPress Dependencies
+ */
+import { withDispatch } from '@wordpress/data';
+
```
👌 2b6f4480b

noisysocks committed:
Don't allow Status & Visibility panel to be hidden

noisysocks committed:
Use isEmptyRender() instead of a snapshot test

noisysocks committed:
Introduce toggleOption() E2E helper

noisysocks commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -0,0 +1,73 @@
+						/>
+					) }
+				/>
+				<EnablePanelOption label={ __( 'Featured Image' ) } panelName="featured-image" />
```
🤔🤔🤔 

OK, so currently we have:

| Component | Identifier |
| --- | --- |
| `PostStatus` | `post-status` |
| `PostTaxonomies` | `taxonomy-panel-*` |
| `FeaturedImage` | `featured-image` |
| `PostExcerpt` | `post-excerpt` |
| `DiscussionPanel` | `discussion-panel` |
| `PageAttributes` | `page-attributes` |

Here's my proposal:

| Component | Identifier |
| --- | --- |
| `StatusPanel` | `status` |
| `PostTaxonomiesPanel` | `post-taxonomy-*` |
| `FeaturedImagePanel` | `featured-image` |
| `PostExcerptPanel` | `post-excerpt` |
| `DiscussionPanel` | `discussion` |
| `PageAttributesPanel` | `page-attributes` |

These names ensure:

- Identifiers consistently don't contain the word `panel`.
- Component names end with `Panel` which properly describes what the UI component actually _is_.
- Panels that only apply to posts begin with `post-`.
- Panels that only apply to pages begin with `page-`.

What do you think? Is it worth changing the names? It's tricky to do this and ensure backwards compatibility with the stored `preferences`.

gziolo replied:
When you check closely code of the panels, it seems like all of them but `PageAttributesPanel` should be prefixed with `Post`. I also agree that all of components should end with `Panel`. My vote would be for:

| Component | Identifier |
| --- | --- |
| `PostStatusPanel` | `post-status` |
| `PostTaxonomiesPanel` | `post-taxonomy-*` |
| `PostFeaturedImagePanel` | `post-featured-image` |
| `PostExcerptPanel` | `post-excerpt` |
| `PostDiscussionPanel` | `post-discussion` |
| `PageAttributesPanel` | `page-attributes` |

What do you think?

gziolo replied:
By the way, this can be done in a seperate PR, this isn't a blocker at all as it existed before. It would be nice to have it standardized as we will be referencing them in the modal dialog. By they way, do you think we should put those values in constants or would it be overkill?

noisysocks replied:
> By they way, do you think we should put those values in constants or would it be overkill?

It's probably overkill. We don't use constants for other identifiers e.g. block names and modal names.

> When you check closely code of the panels, it seems like all of them but PageAttributesPanel should be prefixed with Post.

I disagree. I think:

- Panels that appear on both posts and pages (e.g. Discussion) should have no prefix
- Panels that appear on posts but not pages (e.g. `PostExcept`) should have a `Post` prefix
- Panels that appear on pages but not posts (e.g. `PageAttributes`) should have a `Page` prefix

gziolo replied:
> Panels that appear on both posts and pages (e.g. Discussion) should have no prefix

I like your reasoning. I think it's more complex in the WordPress world because everything is a post both `Post` and `Page` which makes it all even more confusing.

I'm wondering how Classic Editor handles it.

noisysocks committed:
Add Page Attributes screen option

- Allow users to enable or disable Page Attributes within the Options modal.
- Only display the Page Attributes and Post Excerpt options when they
  are applicable.

noisysocks committed:
Make <labels> full width in Options modal

Expands the click area of an option to fill the width of the modal so
that it's easier to toggle on and off.

noisysocks commented:
> 2. Options were moved to their own group which doesn't have a title, should it remain in one of the existing groups instead?

I think the current behaviour makes sense. _Options_ is a "thing" that belongs to the editor. It's not a tool, nor an editor mode, nor a view mode.

> 3. You can toggle options by clicking checkbox or text, in my opinion we should expand it to the full width of the row

👌 Done in c0d7a6f87.

> When using only the keyboard and the modal closes, focus must return to the "invoking context", that is the control that opened the modal. This is an accessibility requirement for modals and it's how the Gutenberg modal is supposed to work. It has been built do have this behavior out of the box.

Good catch! You can see this bug already in `master` by opening the _Keyboard Shortcuts_ modal using your keyboard, pressing <kbd>Esc</kbd>, and then pressing <kbd>Tab</kbd>. The bug is closely related (the same, I daresay) to https://github.com/WordPress/gutenberg/issues/1918. Do you think it's a blocker for this PR?

> The only way I see (so far) to avoid the focus loss is to implement a fallback mechanism in the modal component, allowing to specify a fallback element where to move focus back to.

This is similar to @aduth's suggestion here: https://github.com/WordPress/gutenberg/issues/1918#issuecomment-315770851. I'll have to dwell on this idea a little, as my first instinct is that it would be difficult to maintain these selectors if/when they change.

gziolo commented:
I see you have already fixed the issue with Status & Visibility.

I see that it's possible to hide `Author` fiel. @mtias should we also add it to the modal here?

![screen shot 2018-10-16 at 11 48 46](https://user-images.githubusercontent.com/699132/47007895-820c8600-d139-11e8-97d0-8d7638e70c68.png)

gziolo reviewed (approved):
All my feedback was addressed. It works as expected. There is still this open discussion about doing refactor for panel names and their components: https://github.com/WordPress/gutenberg/pull/10215/files#r225102139. However it isn't a blocker in my opinion. We can handle it as follow-up when we agree how to make it more conistent.

There is still this issue @afercia raised which isn't introduced by this PR. It existed before with the `Keyboard shortcuts` modal dialog, as well with sidebars as filed in #1918. I personally believe, that we should find a proper way of handling this issue globally rather than trying to patch it with some hardcoded focus return in this PR. I added comment https://github.com/WordPress/gutenberg/issues/1918#issuecomment-430177248 to ensure it's taken into account when #1918 is solved.

mtias commented:
> I personally believe, that we should find a proper way of handling this issue globally rather than trying to patch it with some hardcoded focus return in this PR

Should definitely be handled globally as we cannot assume plugin / block authors would adhere to it consistently. I propose keeping track of a `fallbackFocus` reference across the app.

tofumatt commented:
I agree; the issue of focus is not specific to this PR and should not hold it up.

gziolo commented:
Good job @aduth, your check works, Travis fails because:
> There are local uncommitted changes after one or both of 'npm install' or 'npm run docs:build'!

aduth commented:
>It took me a while to figure out because we run everything concurrently.

Yeah, was raised in #core-restapi as well yesterday. See #10625.

afercia commented:
> I agree; the issue of focus is not specific to this PR and should not hold it up.

@tofumatt has a specific issue been opened? If so, can you please me point me at it?
Just noticed it happens also when opening the "Keyboard Shortcuts" modal from the More menu:

<img width="343" alt="screen shot 2018-10-16 at 21 57 40" src="https://user-images.githubusercontent.com/1682452/47043730-a9417280-d18e-11e8-897e-0a05b9f781c5.png">

noisysocks committed:
Regenerate docs

noisysocks commented:
> It took me a while to figure out because we run everything concurrently.

It's pretty hard to spot the error among the noise—thanks for the pointer! 😅

noisysocks commented:
> @tofumatt has a specific issue been opened? If so, can you please me point me at it?

It's the same bug as described in https://github.com/WordPress/gutenberg/issues/1918. @gziolo and I have noted that there are some additional action items to take over in the comments there.

StaggerLeee commented:
Why not add in Options modal an option for resizing editor width ? With live preview.
Some number slider, or number field.

You would make life easier for all non coders who install finished themes. Not a small number, you have to agree.

I personaly each time I go to the Gutenberg to test new things think, damn how narrow is editor content width. I literally stare in huge empty white space.

Someone here on Github mentioned survey, and how this width is most natural for majority of Users. I would like to see results of this survey, not to blindly trust anyone's statement.

noisysocks commented:
@StaggerLeee: We won't be adding any more features to Gutenberg so that we can focus on polishing what we have for WordPress 5.0. I'd encourage you to open up a new issue with your feature suggestion and we can look into it for a future release 😄

StaggerLeee commented:
If you promise me not to close it directly. I am tired of playing Pinata.  😃
